contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,594 @@
+multi-tenancy model) have no mechanism for running on top of Kubernetes without recompilation of the core or
+for extending the core with additional policy. This prevents the formation of an open ecosystem for tools
+*around* Kubernetes, forcing all changes to policy to go through the Kubernetes codebase review gate (when
+such review is unnecessary or disruptive to Kubernetes itself).
+
+
+## Design
+
+It should be possible to perform holistic policy enforcement in Kubernetes without the recompilation of the
+core project as plugins that can be added and removed to a stock Kubernetes release. That extension
+of admission control should leverage similar mechanisms to our existing controller frameworks where possible
+and otherwise be performant and reliable.
+
+
+### Requirements
+
+1.  Easy Initialization
+
+    Privileged components should be able to easily participate in the **initialization** of a new object.
+
+2.  Synchronous Validation
+
+    Synchronous rejection of initialized objects or mutations must be possible outside of the kube-apiserver binary
+
+3.  Backwards Compatible
+
+    Existing API clients must see no change in behavior to external admission other than increased latency
+
+4.  Easy Installation
+
+    Administrators should be able to easily write a new admission plugin and deploy it in the cluster
+
+5.  Performant
+
+    External admission must not significantly regress performance in large and dense clusters
+
+6.  Reliable
+
+    External admission should be capable of being "production-grade" for deployment in an extremely large and dense cluster
+
+7.  Internally Consistent
+
+    Developing an admission controller should reuse as much infrastructure and tools as possible from building custom controllers so as to reduce the cost of extension.
+
+
+### Specification
+
+Based on observation of the actual admission control implementations the majority of mutation
+occurs as part of creation, and a large chunk of the remaining controllers are for side-effect free
+validation of creation and updates. Therefore we propose the following changes to Kubernetes:
+
+1.  Allow some controllers to act as "initializers" - watching the API and mutating the object before it is visible to normal clients.
+
+    This would reuse the majority of the infrastructure in place for controllers. Because creation is
+    one-way, the object can be "revealed" to regular clients once a set list of initializers is consumed. These
+    controllers could run on the cluster as pods. Because initialization is a non-backwards compatible API change,
+    some care must be taken to shield old clients from observing the scenario.
+
+2.  Add a generic **external admission webhook** controller that is non-mutating (thus parallelizable)
+
+    This generic webhook API would resemble `admission.Interface` and be given the input object (for create) and the
```

Comment:
Yeah, I wasn't asking to pass the patch itself, that can be a later addition only if a pitchfork-wielding hoard demands it IMO. Might be worth calling out that patch application may result in many calls to the admission controller.